File: products/connection-hub/packages/connection-hub/src/connection_hub/operation_groups.py
"""Operation groups a service declares for a Card editor (W260).

A service catalog names, beside its operations, how they are grouped (for
Problem Board: Review, Work, Plan, People). Each operation carries ``group``;
the resource or namespace carries ``operation_groups``, a map of group key to
``{label, order}`` (a bare string is the label). Connection Hub's Card view
groups by it, so no client builds its own grouping. Presentation only: it
stays out of every descriptor digest, so regrouping never raises drift.
"""

from __future__ import annotations

from typing import Any, Mapping


def _text(value: Any) -> str:
    return str(value or "").strip()


def parse_operation_groups(raw: Any) -> tuple[dict[str, Any], ...]:
    """The declared groups in display order: by ``order``, then declaration order."""

    if not isinstance(raw, Mapping):
        return ()
    rows: list[tuple[float, int, dict[str, Any]]] = []
    for index, (key, value) in enumerate(raw.items()):
        name = _text(key)
        if not name:
            continue
        data = value if isinstance(value, Mapping) else {"label": value}
        try:
            order = float(data["order"]) if data.get("order") is not None else float(index)
        except (TypeError, ValueError):
            order = float(index)
        rows.append((order, index, {"group": name, "label": _text(data.get("label")) or name, "order": order}))
    return tuple(row for _order, _index, row in sorted(rows, key=lambda item: (item[0], item[1])))


def _drop(entry: Any, key: str) -> Any:
    if not isinstance(entry, Mapping):
        return entry
    return {name: value for name, value in entry.items() if name != key}


def without_grouping(named_services: Any) -> Any:
    """A ``named_services`` tree without its grouping, for a descriptor digest.

    Only where the schema puts grouping: ``operation_groups`` on a namespace,
    ``group`` on a tool entry and on an operation entry. A tool, operation or
    namespace that is itself named ``group`` is content and stays.
    """

    if not isinstance(named_services, Mapping):
        return named_services
    namespaces = named_services.get("namespaces")
    if not isinstance(namespaces, Mapping):
        return dict(named_services)
    stripped: dict[str, Any] = {}
    for name, namespace in namespaces.items():
        namespace = _drop(namespace, "operation_groups")
        tools = namespace.get("tools") if isinstance(namespace, Mapping) else None
        if isinstance(tools, Mapping):
            kept_tools: dict[str, Any] = {}
            for tool_name, tool in tools.items():
                tool = _drop(tool, "group")
                operations = tool.get("operations") if isinstance(tool, Mapping) else None
                if isinstance(operations, Mapping):
                    tool = {
                        **tool,
                        "operations": {op: _drop(policy, "group") for op, policy in operations.items()},
                    }
                kept_tools[tool_name] = tool
            namespace = {**namespace, "tools": kept_tools}
        stripped[name] = namespace
    return {**named_services, "namespaces": stripped}

File: products/connection-hub/apps/connection-hub@1-0/ui/widgets/connections/src/features/delegatedAccess/operationGroups.ts
// W260 (operator ruling, 2026-09-26): a service declares how its operations are
// grouped (for Problem Board: Review, Work, Plan, People) in its catalog, and
// the Card editor groups by it. No client builds its own grouping.
//
// Each operation may carry `group`; the resource or namespace carries
// `operation_groups`, already sorted by the server ({group, label, order}).
// Operations keep their catalog order inside a group. An operation without a
// declared group, or naming a group the map omits, lands in its own group
// labelled by the key, or in "Other" when it has none.

export interface OperationGroupDeclaration {
  group: string;
  label?: string;
  order?: number;
}

export interface OperationGroup<T> {
  key: string;
  label: string;
  items: T[];
}

export const OTHER_GROUP_LABEL = 'Other';

/** The operations under their declared groups, or null when the service declares none. */
export function groupOperations<T>(
  items: readonly T[],
  groupOf: (item: T) => string | undefined,
  declared: readonly OperationGroupDeclaration[] | undefined,
): OperationGroup<T>[] | null {
  const anyGrouped = items.some((item) => Boolean((groupOf(item) || '').trim()));
  if (!anyGrouped) return null;
  const groups: OperationGroup<T>[] = [];
  const byKey = new Map<string, OperationGroup<T>>();
  (declared || []).forEach((row) => {
    const key = (row.group || '').trim();
    if (!key || byKey.has(key)) return;
    const group = { key, label: (row.label || '').trim() || key, items: [] as T[] };
    byKey.set(key, group);
    groups.push(group);
  });
  const extra: OperationGroup<T>[] = [];
  let other: OperationGroup<T> | null = null;
  items.forEach((item) => {
    const key = (groupOf(item) || '').trim();
    if (!key) {
      other = other || { key: '', label: OTHER_GROUP_LABEL, items: [] };
      other.items.push(item);
      return;
    }
    let group = byKey.get(key);
    if (!group) {
      group = { key, label: key, items: [] };
      byKey.set(key, group);
      extra.push(group);
    }
    group.items.push(item);
  });
  return [...groups, ...extra, ...(other ? [other] : [])].filter((group) => group.items.length > 0);
}
